### ISSUE TITLE
Process to reconnect to remote hosts broken - "Remote - SSH"

### ISSUE THREAD

**Issue by umdstu** [microsoft/vscode-remote-release#5231]
I have searched for existing similar issues, and with the exceptions of the ones i will link to below, all appear to be specifically related to WSL. I am not using WSL however. The behavior is that upon reconnecting to the localhost (AWS), all of the vscode windows have disconnected from the put to sleep remote host/container (which is fine), the problem is reconnecting. There used to be a very nice reconnect dialog (once I set the `window.dialogStyle` to `custom`), but then that stopped working in v53. So this has been going on sporadically (but more recently it seems from the last 1 or 2 updates) ever since the team attempted to fix/clarify some things in [v53](https://github.com/microsoft/vscode-docs/blob/main/remote-release-notes/v1_53.md#less-disruptive-disconnects). @roblourens thought things were fixed with https://github.com/microsoft/vscode-remote-release/issues/4441 via https://github.com/microsoft/vscode-remote-release/issues/4389, but it's never been completely fixed for me (or it's unrelated to that after all). https://github.com/microsoft/vscode-remote-release/issues/4529 May be related, but it was opened the day before

One of the odd behaviors that I was experiencing prior to setting `windowStyle` to `custom`, was that I could not select the dialog buttons in all but one of the vscode windows. That seems to be sporadically happening again (despite the setting still being in place), not every single time.

To summarize, the behavior that has been happening include:

- Unable to click reconnect buttons in reconnect dialog
- Clicking the reconnect button causes all open vscode windows to close (and not reopen)
- I think there is a third but it is alluding my memory on this fine morning, will update when I remember or re-experience it this week, but hopefully this is enough to get started.

Steps to Reproduce:

1. Start up vscode with multiple windows open and a remote ssh connection for each.
2. Let the local host be shutdown (For me, it happens overnight, when the local host (AWS) has been put to sleep, therefore the remote host is shutdown as well)
3. Upon logging into the local host, all vscode windows should be disconnected (expected)
4. For each window, click the button to reconnect your session
5. Experience one of the 3 scenarios randomly:
    1. Window reconnects correctly, or
    2. VS Code completely exits, or
    3. Reconnect button is not clickable

Thanks!

- VSCode Version: 1.57.1
- Local OS Version: Amazon Linux 2
- Remote OS Version: CentOS 7 / Docker
- Remote Extension/Connection Type: SSH
- Remote Extension Version: 0.65.7
- Logs: Are seemingly blown away by this process

**Comment by IssamLaradji** [comment]
I keep getting this as well! with this message

![image](https://user-images.githubusercontent.com/3382128/122776867-7c1bbb00-d279-11eb-8c22-1f6a89d65151.png)

**Comment by bamurtaugh** [comment]
@umdstu Thanks for the detailed report. Do you see the same "Error: Handler already set!" message taht @IssamLaradji shows? 

@IssamLaradji Do you have any logs from the above scenario, or do yours also get blown away? https://code.visualstudio.com/docs/remote/troubleshooting#_reporting-issues

**Comment by umdstu** [comment]
I do not recall ever seeing that pop-up, sorry.

Let me know if there’s anything I can enable or do to help. I’ll also try
to trigger it if I can!

**Comment by IssamLaradji** [comment]
how do I get the logs for this? this happens for specific folders where the `ssh` connection got broken due to bad connection. Other ssh folders work fine. But once this issue happens for a specific folder, it never recovers for that folder and that error happens all the time. Is there a way to clear the ssh cache?

**Comment by bamurtaugh** [comment]
@IssamLaradji The link provided describes how to gather logs: https://code.visualstudio.com/docs/remote/troubleshooting#_reporting-issues (i.e. **Remote-SSH: Show Log**).

I think it'd be worth it for you to try **Remote-SSH: Kill VS Code Server on Host** - this will remove the server.

**Comment by IssamLaradji** [comment]
@bamurtaugh that seemed to work yes, thanks! Sorry I removed my previous message because I didn't want to steal the focus from @umdstu 's issue.

**Comment by IssamLaradji** [comment]
nvm the issue is happening again, and I can't get the logs because it just freezes
![image](https://user-images.githubusercontent.com/3382128/124005043-a4589780-d9a6-11eb-9076-f40629e14e95.png)

**Comment by roblourens** [comment]
> One of the odd behaviors that I was experiencing prior to setting windowStyle to custom, was that I could not select the dialog buttons in all but one of the vscode windows.

This is an issue with some window managers on linux. We show multiple native dialogs and only one of them is on top and clickable. We can't do anything about that besides have you switch to custom dialogs, and it sounds like you found that already.

> Clicking the reconnect button causes all open vscode windows to close (and not reopen)

No idea what that is about. So you click the "Reload Window" button, right? And the window just exits?

If you can set it up to happen again, maybe you could start `code --verbose` in a terminal, repro the issue, then see if there is some useful crash log in the terminal

**Comment by umdstu** [comment]
@roblourens Sorry for the delay. I kept forgetting to run vscode in the terminal so I had something to capture for you. All last week I remembered to do it, and so this morning we finally have something!

This is in regards to the "Reconnect" leading to vscode completely exiting. Here is the first error so I'll share this.

```console
[27585:1122/051727.849321:INFO:CONSOLE(602)] "%c  ERR color: #f33 Unable to read file 'vscode-remote://ssh-remote+sonicdev/sonicdev/workspace/infrastructure/ci-templates/scripts/version-checker.py' (Canceled: Canceled): Error: Unable to read file 'vscode-remote://ssh-remote+sonicdev/sonicdev/workspace/infrastructure/ci-templates/scripts/version-checker.py' (Canceled: Canceled)
    at yi.doReadFileStream (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:602:27791)
    at async f.doRead (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2388:42320)
    at async f.readStream (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2388:42041)
    at async tn.resolveFromFile (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2377:98223)", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (602)
[main 2021-11-22T10:17:28.064Z] menubarService#updateMenubar 1
[27585:1122/051749.247574:INFO:CONSOLE(29)] "%cTRACE color: #888 Started scanning user extensions", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js (29)
[27585:1122/051749.279346:INFO:CONSOLE(29)] "%cTRACE color: #888 Scanned user extensions: 24", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js (29)
[27585:1122/051949.500868:INFO:CONSOLE(29)] "%cTRACE color: #888 ExtensionManagementService.refreshReportedCache", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js (29)
[27585:1122/051949.501093:INFO:CONSOLE(29)] "%cTRACE color: #888 RequestService#request https://az764295.vo.msecnd.net/extensions/marketplace.json", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js (29)
[27585:1122/051949.508951:INFO:CONSOLE(29)] "%cTRACE color: #888 ExtensionManagementService.refreshReportedCache - got 6 reported extensions from service", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js (29)
[27585:1122/051949.510311:INFO:CONSOLE(29)] "%cTRACE color: #888 Started scanning user extensions", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js (29)
[27585:1122/051949.577377:INFO:CONSOLE(29)] "%cTRACE color: #888 Scanned user extensions: 24", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js (29)
[27585:1122/052202.862078:INFO:CONSOLE(602)] "%cTRACE color: #888 [text file model] resolve() - enter vscode-remote://ssh-remote+sonicdev/sonicdev/workspace/infrastructure/ci-templates/scripts/version-checker.py", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (602)
[27585:1122/052202.862962:INFO:CONSOLE(602)] "%cTRACE color: #888 [text file model] resolveFromFile() vscode-remote://ssh-remote+sonicdev/sonicdev/workspace/infrastructure/ci-templates/scripts/version-checker.py", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (602)
[27585:1122/052202.876411:INFO:CONSOLE(602)] "%c  ERR color: #f33 Unable to read file 'vscode-remote://ssh-remote+sonicdev/sonicdev/workspace/infrastructure/ci-templates/scripts/version-checker.py' (Canceled: Canceled): Error: Unable to read file 'vscode-remote://ssh-remote+sonicdev/sonicdev/workspace/infrastructure/ci-templates/scripts/version-checker.py' (Canceled: Canceled)
    at yi.doReadFileStream (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:602:27791)
    at async f.doRead (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2388:42320)
    at async f.readStream (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2388:42041)
    at async tn.resolveFromFile (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2377:98223)", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (602)
[27585:1122/052202.912534:INFO:CONSOLE(602)] "%cTRACE color: #888 CommandService#executeCommand workbench.action.reloadWindow", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (602)
[main 2021-11-22T10:22:02.913Z] Lifecycle#unload() - window ID 1
[27585:1122/052202.941785:INFO:CONSOLE(602)] "%cTRACE color: #888 [lifecycle] onBeforeUnload (reason: 3)", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (602)
[27585:1122/052202.956703:INFO:CONSOLE(602)] "%cTRACE color: #888 [lifecycle] onBeforeUnload continues without veto", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (602)
[main 2021-11-22T10:22:02.957Z] Lifecycle#unload() - no veto (window ID 1)
[27585:1122/052202.960805:INFO:CONSOLE(602)] "%cTRACE color: #888 [lifecycle] onWillUnload (reason: 3)", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (602)
[main 2021-11-22T10:22:03.025Z] StorageMainService: closed workspace storage (cc8bfc892581c4d6c4e87f1e2c339f7d)
/dev/fd/3: No such file or directory
Server response:
/dev/fd/3: No such file or directory
Server response:
```

I hope this helps!

**Comment by roblourens** [comment]
@deepak1556 any idea why vscode seems to be quitting when we try to reload? I have no idea where the last few log lines come from

**Comment by roblourens** [comment]
Also can you try downloading the Insiders' build and tell me whether it happens there? https://code.visualstudio.com/insiders

**Comment by deepak1556** [comment]
```
/dev/fd/3: No such file or directory
Server response:
/dev/fd/3: No such file or directory
Server response:
```

Indicates there is a crash in one of the client processes. Also the original report from the user with vscode `1.57` had broken support on RHEL distros due to https://github.com/microsoft/vscode/issues/115784, @umdstu can you check if the issue happens with latest stable or insiders as @roblourens suggested ?

**Comment by umdstu** [comment]
Thanks for opening @roblourens 

For what it's worth, this crash happened under 1.62.3, which appears to be the latest. This system auto-updates so it should always have the latest. However, it's happened sporadically over the past 6-9mo (maybe more), so I don't think it's version dependent. But yes, remote VM is RHEL 7. I will download latest and run it, but it could take some time to see if it fixes anything. I will report back (could be some time) when it happens again, unless you have more questions.

As always, thanks for the help.

**Comment by deepak1556** [comment]
If you are able to repro the crash with `1.62.3`, can you attach the crash dumps following the steps at https://github.com/microsoft/vscode/wiki/Native-Crash-Issues

**Comment by umdstu** [comment]
Unable to install `-insiders` for the time being, but I did kick off vscode with the `--crash-reporter-directory` option so ill report back next time it happens.

**Comment by umdstu** [comment]
Good morning,

So this morning after re-establishing the SSH tunnel to my remote container, and trying to 'reconnect' to it via the vscode prompt to do so, all of the vscode windows closed again. I went to check the crash report directory that I had specified some time ago, and it's essentially empty. There is only one file. in `exthost Crash Reports/` called `client_id` and in that file is a single UUID, nothing else. Per the link above I expected to see a *.dmp file but I do not.

I'm sorry that I couldn't provide anything.
